algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/spectrum_augmenter.py

```diff
@@ -0,0 +1,160 @@
+    diagonal = torch.arange(mask_size, device=device).reshape(1, 1, -1)
+    diagonal = torch.tile(diagonal, [batch_size, multiplicity, 1])
+    pre_mask = torch.minimum(diagonal < end, diagonal > start)
+
+    # Sum masks with appropriate multiplicity.
+    if masks_per_frame > 0:
+      multiplicity_weights = torch.tile(
+          torch.arange(multiplicity, device=device).long()[None, ...],
+          [batch_size, 1])
+      multiplicity_tensor = masks_per_frame * choose_range
+      multiplicity_weights = (multiplicity_weights < multiplicity_tensor).long()
+      pre_mask = torch.einsum('bmt,bm->bt', pre_mask, multiplicity_weights)
+    else:
+      pre_mask = torch.einsum('bmt->bt', pre_mask)
+    mask = 1.0 - (pre_mask > 0).long()
+
+    return mask
+
+  def _time_mask(self, inputs, length):
+    # Get time masking parameters.
+    time_mask_max_frames = self.time_mask_max_frames
+    use_dynamic_time_mask_max_frames = self.use_dynamic_time_mask_max_frames
+    multiplicity = self.time_mask_count
+    max_ratio = self.time_mask_max_ratio
+
+    # If maximum mask length is zero, do nothing.
+    if ((time_mask_max_frames == 0 and not use_dynamic_time_mask_max_frames) or
+        max_ratio <= 0.0):
+      return inputs
+    if multiplicity == 0:
+      return inputs
+    batch_size, time_length, _ = inputs.shape
+
+    # When using dynamic time mask size, discard upper-bound on
+    # maximum allowed frames for time mask.
+    if use_dynamic_time_mask_max_frames:
+      time_mask_max_frames = None
+    # Create masks in time direction and apply.
+    block_arrays = self._get_mask(
+        batch_size,
+        choose_range=length,
+        mask_size=time_length,
+        max_length=time_mask_max_frames,
+        masks_per_frame=self.time_masks_per_frame,
+        multiplicity=multiplicity,
+        max_ratio=max_ratio,
+        device=inputs.device)
+
+    outputs = torch.einsum('bxy,bx->bxy', inputs, block_arrays)
+    return outputs
+
+  def _frequency_mask(self, inputs):
+    # Mask parameters.
+    freq_mask_max_bins = self.freq_mask_max_bins
+    multiplicity = self.freq_mask_count
+
+    # If masking length or count is zero, do nothing.
+    if freq_mask_max_bins == 0 or multiplicity == 0:
+      return inputs
+
+    # Arguments to pass to mask generator.
+    batch_size, _, num_freq = inputs.shape
+    choose_range = num_freq * torch.ones(batch_size, device=inputs.device)
+    # Create masks in frequency direction and apply.
+    block_arrays = self._get_mask(
+        batch_size,
+        choose_range=choose_range,
+        mask_size=num_freq,
+        max_length=freq_mask_max_bins,
+        masks_per_frame=0.0,
+        multiplicity=multiplicity,
+        max_ratio=1.0,
+        device=inputs.device)
+
+    outputs = torch.einsum('bxy,by->bxy', inputs, block_arrays)
+    return outputs
+
+  def forward(self, inputs, paddings):
+    lengths = torch.einsum('bh->b', 1 - paddings).long()
+
```

Comment:
nit: delete unnecessary newline (here and below)

Reply:
Done.